services/apps/src/apps/service/oauth.clj

```diff
@@ -32,3 +36,47 @@
         token-callback (partial op/store-access-token api-name username)]
     (authy/get-access-token (build-authy-server-info server-info token-callback) code)
     {:state_info state-info}))
+
+(defn- format-admin-token-info
+  "Formats access token info for administrative endpoints."
+  [token-info]
+  (when token-info
+    {:access_token  (:access-token token-info)
+     :expires_at    (.. (:expires-at token-info) toInstant toEpochMilli)
+     :refresh_token (:refresh-token token-info)
+     :webapp        (:webapp token-info)}))
+
+(defn- format-token-info
+  "Formats access token info."
+  [token-info]
+  (when token-info
+    (-> (format-admin-token-info token-info)
+        (select-keys [:expires_at :webapp]))))
+
+(defn get-token-info
+  "Retrieves the user's token information, excluding the actual tokens, for an external API if it exists."
+  [api-name {:keys [username]}]
+  (or (format-token-info (op/get-access-token api-name username))
+      (cxu/not-found "access token not found" :api_name api-name)))
+
+(defn get-admin-token-info
+  "Retrieves the user's token information for an external API if it exists."
+  [api-name {:keys [username]}]
+  (or (format-admin-token-info (op/get-access-token api-name username))
+      (cxu/not-found "access token not found" :api_name api-name)))
+
+(defn authorization-uri
+  "Generates an authorization URI for a remote API."
+  [{:keys [api-name] :as server-info} username state-info]
+  (when-not (op/get-access-token api-name username)
+    (str (assoc (curl/url (:auth-uri server-info))
+                :query {:response_type "code"
+                        :client_id     (:client-key server-info)
+                        :redirect_uri  (:redirect-uri server-info)
+                        :state         (op/store-authorization-request username state-info)}))))
+
+(defn get-redirect-uris
+  "Retrieves the redirect URIs that can be used for a user to authenticate to a remote API."
+  [{:keys [username]}]
+  (let [build-auth-uri (fn [[_ server-info-fn]] (authorization-uri (server-info-fn) username ""))]
+    (remove-vals nil? (into {} (map (juxt key build-auth-uri) server-info-fn-for)))))
```

Comment:
I just realized that there's a problem with using `server-info-fn-for` as-is for this endpoint. The problem is that we'll still try to create the redirect URI even if Agave is disabled, which can cause problems. I'll make a commit to ensure that this won't happen momentarily.